python/ql/test/experimental/dataflow/coverage/argumentPassing.py

```diff
@@ -0,0 +1,188 @@
+import sys
+import os
+
+sys.path.append(os.path.dirname(os.path.dirname((__file__))))
+from testlib import *
+
+arg = "source"
+arg1 = "source1"
+arg2 = "source2"
+arg3 = "source3"
+arg4 = "source4"
+arg5 = "source5"
+arg6 = "source6"
+arg7 = "source7"
+
+
+def SINK(x, expected=arg):
+    if x == expected:
+        print("OK")
+    else:
+        print("Unexpected flow", x)
+
+
+def SINK1(x):
+    SINK(x, expected=arg1)
+
+
+def SINK2(x):
+    SINK(x, expected=arg2)
+
+
+def SINK3(x):
+    SINK(x, expected=arg3)
+
+
+def SINK4(x):
+    SINK(x, expected=arg4)
+
+
+def SINK5(x):
+    SINK(x, expected=arg5)
+
+
+def SINK6(x):
+    SINK(x, expected=arg6)
+
+
+def SINK7(x):
+    SINK(x, expected=arg7)
+
+
+def argument_passing(
+    a,
+    b,
+    /,
+    c,
+    d="",
+    *,
+    e="",
+    f,
+    **g,
+):
+    SINK1(a)
+    SINK2(b)
+    SINK3(c)
+    SINK4(d)
+    SINK5(e)
+    SINK6(f)
+    SINK7(g["g"])
+
+
+@expects(7)
+def test_argument_passing():
+    argument_passing(arg1, *(arg2, arg3, arg4), e=arg5, **{"f": arg6, "g": arg7})
+
+
+def with_pos_only(a, /, b):
+    SINK1(a)
+    SINK2(b)
+
+
+@expects(6)
+def test_pos_only():
+    with_pos_only(arg1, arg2)
+    with_pos_only(arg1, b=arg2)
+    with_pos_only(arg1, *(arg2,))
+
+
+def with_multiple_kw_args(a, b, c):
+    SINK1(a)
+    SINK2(b)
+    SINK3(c)
+
+
+@expects(9)
+def test_multiple_kw_args():
+    with_multiple_kw_args(b=arg2, c=arg3, a=arg1)
+    with_multiple_kw_args(arg1, *(arg2,), arg3)
+    with_multiple_kw_args(arg1, **{"c": arg3}, b=arg2)
+
+
+def with_default_arguments(a=arg1, b=arg2, c=arg3):
+    SINK1(a)
+    SINK2(b)
+    SINK3(c)
+
+
+@expects(12)
+def test_default_arguments():
+    with_default_arguments()
+    with_default_arguments(arg1)
+    with_default_arguments(b=arg2)
+    with_default_arguments(**{"c": arg3})
+
+
+# Nested constructor pattern
+def grab_foo_bar_baz(foo, **kwargs):
+    SINK1(foo)
+    grab_bar_baz(**kwargs)
+
+
+def grab_bar_baz(bar, **kwargs):
+    SINK2(bar)
+    try:
+        SINK2(kwargs["bar"])
+    except:
+        print("OK")
```

Comment:
```suggestion
```

this part doesn't make any sense, since you would never be able to pass in `bar` as a part of the **kwargs, since it is a valid keyword argument.

```
def func(bar, **kwargs): 
    print(bar, kwargs)

func(1, bar="bar") # TypeError: func() got multiple values for argument 'bar'
func(bar=1, bar="bar") # SyntaxError: keyword argument 
repeated
func(1, **{"bar": 2}) # TypeError: func() got multiple values for argument 'bar'

Reply:
Indeed. However, without the clearing of content, dataflow will find a path. So this is a negative test.

Reply:
Note that the tests prints `"OK"` in case `"bar"` is not in `kwargs`.

Reply:
Aha :bulb: that makes sense! Could you add a comment to the code explaining this? :blush:

Reply:
I first added a comment, but then I defined `SINK2_F` to make it clear that flow is not expected. Hopefully a comment is now not needed.

Reply:
yes, the `SINK2_F` looks very good :+1: